### ISSUE TITLE
CustomAnalyzerDefinition allows filters which are not handled by AnalysisDefinition

### ISSUE THREAD

**Issue by jeantil** [sksamuel/elastic4s#301]
`com.sksamuel.elastic4s.CustomAnalyzerDefinition` allows any type extending `com.sksamuel.elastic4s.AnalyzerFilter`  as filters and internally finds filters with the following selectors : 

````scala
val tokenFilters = filters.collect { case token: TokenFilter => token }
val charFilters = filters.collect { case token: CharFilter => token }
````

I am therefore able to create my own tokenfilter to pass to a CustomAnalyzerDefinition such as : 
````scala
 case class KeywordSynonymTokenFilter(name: String, path: String, ignoreCase: Boolean = false, expand: Boolean = true)
      extends TokenFilter with AnalyzerFilterDefinition {

    val filterType = "synonym"

    override def build(source: XContentBuilder): Unit = {
      source.field("synonyms_path", path)
      source.field("ignore_case", ignoreCase)
      source.field("expand", expand)
      source.field("tokenizer", "keyword")
    }
  }
````
Passing it to a `CustomAnalyzerDefinition` will indeed compile.

However once the `CustomAnalyzerDefinition` is processed by `com.sksamuel.elastic4s.CreateIndexDefinition`, the selector used are 

````scala
   analyzers.collect {
      case custom: CustomAnalyzerDefinition => custom
    }.flatMap(_.filters).collect {
      case token: TokenFilterDefinition => token
    }

    analyzers.collect {
      case custom: CustomAnalyzerDefinition => custom
    }.flatMap(_.filters).collect {
      case char: CharFilterDefinition => char
    }
````
where `TokenFilterDefinition` and `CharFilterDefinition` both are sealed traits which only seem to be used as tag types for `TokenFilter with AnalyzerFilterDefinition` and `CharFilter with AnalyzerFilterDefinition`

This leads to the possibility of creating "invalid" `CustomAnalyzerDefinition` which will generate runtime elasticsearch errors when run through `CreateIndexDefinition` since the filters referenced by the analyzer will not be properly created. 

I discovered this while trying to work around #300

**Comment by sksamuel** [comment]
So it seems simply changing the collects to use the trait should be enough.

**Comment by jeantil** [comment]
Here is how I changed the collects to work around this locally : 
````scala
case class AnalysisDefinition(analyzers: Iterable[AnalyzerDefinition]) {

  def tokenizers: Iterable[Tokenizer] =
    analyzers.collect {
      case custom: CustomAnalyzerDefinition => custom
    }.map(_.tokenizer).filter(_.customized)

  def tokenFilterDefinitions: Iterable[TokenFilter with AnalyzerFilterDefinition] =
    analyzers.collect {
      case custom: CustomAnalyzerDefinition => custom
    }.flatMap(_.filters).collect {
      case token: TokenFilter with AnalyzerFilterDefinition => token
    }

  def charFilterDefinitions: Iterable[CharFilter with AnalyzerFilterDefinition] =
    analyzers.collect {
      case custom: CustomAnalyzerDefinition => custom
    }.flatMap(_.filters).collect {
      case char: CharFilter with AnalyzerFilterDefinition => char
    }
}
````

I am wondering why the TokenFilter trait is sealed and if maybe it could simply be replaced by a type alias or a normal trait.

**Comment by sksamuel** [comment]
I think it makes sense to merge TokenFilterDefinition and TokenFilter. And same for char filters. 
I'm not sure why we have the dual traits, perhaps one was intended for definitions provided by the library, but if you want to use them, you have to use the *Definition ones, which are sealed. @fehmicansaglam wdyt ?

**Comment by sksamuel** [comment]
Ok so the trick now, is that your custom token / char filters should extend CharFilterDefinition and TokenFilterDefinition (better names needed perhaps). These are the ones that are customisable (in the sense that you provide a body that is registered) verus CharFilter / TokenFilter which are used for built in filters. Can you test from master?

**Comment by jeantil** [comment]
Hello 

just tested from master and it works just fine for me. 

Thanks

**Comment by sksamuel** [comment]
This will be included in 1.5.6.

**Comment by jeantil** [comment]
Thanks !

Le mar. 28 avr. 2015 19:47, Stephen Samuel <notifications@github.com> a
écrit :

> This will be included in 1.5.6.
